Add RAdam capturable API for forloop

janeyx99 opened the pull request:
Implementation thanks to @MarouaneMaatouk in https://github.com/pytorch/pytorch/pull/118697, though I've since cleaned it up a lot to save perf on the rect < 5 eager case. It also just looks better now :) Added tests and the cudagraph health check.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121269
* #121264
* __->__ #121260



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

janeyx99 committed:
Add RAdam capturable API for forloop

[ghstack-poisoned]

janeyx99 committed:
Update on "Add RAdam capturable API for forloop"

Implementation thanks to MarouaneMaatouk in https://github.com/pytorch/pytorch/pull/118697. Added tests and the cudagraph health check.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

janeyx99 committed:
Update on "Add RAdam capturable API for forloop"

Implementation thanks to MarouaneMaatouk in https://github.com/pytorch/pytorch/pull/118697. Added tests and the cudagraph health check.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

janeyx99 commented on test/test_cuda.py:
```diff
@@ -2713,10 +2713,6 @@ def test_graph_optims(self):
             (optimizer_ctor, {"lr": 0.1, "foreach": True, "maximize": maximize, "weight_decay": weight_decay})
             for optimizer_ctor, maximize, weight_decay in product((torch.optim.ASGD,), (False, True), (0, 0.1))
-        ] + [
-            (torch.optim.RAdam, {"lr": 0.1, "foreach": True, "decoupled_weight_decay": decoupled_weight_decay,
```
RAdam is included with NAdam on 2701. A great followup is to move this to also use optim info so this logic can be deprecated :p

mlazos reviewed (approved):
Looks good, I think Alban's comment about the perf of computing both branches should be addressed though.

janeyx99 committed:
Update on "Add RAdam capturable API for forloop"

Implementation thanks to MarouaneMaatouk in https://github.com/pytorch/pytorch/pull/118697. Added tests and the cudagraph health check.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
